examples/pettingzoo/meltingpot_env.py

```diff
@@ -0,0 +1,149 @@
+""" PettingZoo interface to meltingpot environments"""
+
+import dm_env
+from meltingpot.python import substrate
+from ml_collections import config_dict
+from gym import spaces
+import numpy as np
+from functools import lru_cache
+
+from gym.utils import EzPickle
+from pettingzoo.utils import wrappers
+from pettingzoo.utils.conversions import parallel_to_aec_wrapper
+from pettingzoo.utils.env import ParallelEnv
+import tree
+import matplotlib.pyplot as plt
+
+PLAYER_STR_FORMAT = 'player_{index}'
+MAX_CYCLES = 1000
+
+def _timestep_to_observations(timestep: dm_env.TimeStep):
+  gym_observations = {}
+  for index, observation in enumerate(timestep.observation):
+    gym_observations[PLAYER_STR_FORMAT.format(index=index)] = {
+        key: value for key, value in observation.items() if 'WORLD' not in key
+    }
+  return gym_observations
+
+
+def _remove_world_observations_from_space(
+    observation: spaces.Dict) -> spaces.Dict:
+  return spaces.Dict({
+      key: observation[key] for key in observation if 'WORLD' not in key})
+
+
+def _spec_to_space(spec: tree.Structure[dm_env.specs.Array]) -> spaces.Space:
+  """Converts a dm_env nested structure of specs to a Gym Space.
+
+  BoundedArray is converted to Box Gym spaces. DiscreteArray is converted to
+  Discrete Gym spaces. Using Tuple and Dict spaces recursively as needed.
+
+  Args:
+    spec: The nested structure of specs
+
+  Returns:
+    The Gym space corresponding to the given spec.
+  """
+  if isinstance(spec, dm_env.specs.DiscreteArray):
+    return spaces.Discrete(spec.num_values)
+  elif isinstance(spec, dm_env.specs.BoundedArray):
+    return spaces.Box(spec.minimum, spec.maximum, spec.shape, spec.dtype)
+  elif isinstance(spec, dm_env.specs.Array):
+    if np.issubdtype(spec.dtype, np.floating):
+      return spaces.Box(-np.inf, np.inf, spec.shape, spec.dtype)
+    elif np.issubdtype(spec.dtype, np.integer):
+      info = np.iinfo(spec.dtype)
+      return spaces.Box(info.min, info.max, spec.shape, spec.dtype)
+    else:
+      raise NotImplementedError(f'Unsupported dtype {spec.dtype}')
+  elif isinstance(spec, (list, tuple)):
+    return spaces.Tuple([_spec_to_space(s) for s in spec])
+  elif isinstance(spec, dict):
+    return spaces.Dict({key: _spec_to_space(s) for key, s in spec.items()})
+  else:
+    raise ValueError('Unexpected spec: {}'.format(spec))
+
+def parallel_env(env_config, max_cycles=MAX_CYCLES):
```

Comment:
This seems to be the only factory used in the training file. I suggest making only this one public, and calling it `get_parallel_env` instead. It should also have a docstring explaining what it does, and what is expected to be the properties of the returned object (also add the PyType annotation for the return).